### PR TITLE
QKV parameters unfused path fixes and optimization

### PR DESCRIPTION
**Pull request by ksivaman**
This PR optimizes the case when the QKV parameters are unfused by removing the need to concatenate and split during fprop and bprop respectively. Essentially, this is achieved by keeping the full W&B buffer in contiguous memory and indexing chunks from it to create the split parameters. The `NoopCat` autograd function is necessary to correctly propagate gradients during backward phase. This design has the following advantages:

1. The weights are owned by the actual TE modules and not `TransformerLayer` (or user).
2. Backwards compatible with previous system where we used the `skip_weight_param_allocation` argument to achieve QKV un-fused.
3. The `wandb_param_split` arg allows splitting the params into any arbitrary number along the first dim; not just 3 for Q, K, V. E.g. in cross attention, we can use a 2 way split to achieve fused key-value layer.

**Note:** This drastically improves BERT phase-1 training convergence, indicating a bug (not found yet) in the previous method of handling this case.

### CODE REVIEW

File: transformer_engine/pytorch/module.py [hunk 1]
```diff
@@ -135,6 +135,44 @@ def _prepare_backward(fp8: bool,
...
+    @staticmethod
+    def forward(ctx, full_param_buffer, *params_split):
+        assert not full_param_buffer.requires_grad, "Buffers should not require gradient"
+        assert (
+            full_param_buffer.shape[0] % len(params_split) == 0
+        ), "Dimensions not compatible for concatenation"
+
+        split_size = full_param_buffer.shape[0] // len(params_split)
+        for i, p in enumerate(params_split):
+            p.data = full_param_buffer[i * split_size : (i+1) * split_size]
```
Comment (ptrendx) [inline transformer_engine/pytorch/module.py:150]: I don't think I follow the logic here - shouldn't that be a check instead?

As in, I give the set of buffers to this op to be concatenated (and the expectation is that they already are concatenated in memory), so why am I setting their pointers to something potentially different?

File: transformer_engine/pytorch/module.py [hunk 2]
```diff
@@ -135,6 +135,44 @@ def _prepare_backward(fp8: bool,
...
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        full_param_buffer, *params_split = ctx.saved_tensors
+
+        split_size = full_param_buffer.shape[0] // len(params_split)
+        grads = []
+
+        for i, p in enumerate(params_split):
+            p.data = full_param_buffer[i * split_size : (i+1) * split_size]
```
Comment (ptrendx) [inline transformer_engine/pytorch/module.py:170]: This is the same as the logic in forward? Why?

Comment (ksivaman) [reply transformer_engine/pytorch/module.py:170]: You're right in the sense that for our use case it doesn't matter since we initialize the individual params from the full buffer's memory. Having this just keeps the `NoopCat` function more general and doesn't assume the params are already correctly populated, but if we don't care about that then we can remove it.

Comment (ptrendx) [reply transformer_engine/pytorch/module.py:170]: Hm? But it is not what that does - it just sets the pointers, not actually copying the data?

Comment (ksivaman) [reply transformer_engine/pytorch/module.py:170]: For correct functionality of fwd/bwd/checkpointing that's all we need right, explicit copies aren't required

Comment (ptrendx) [reply transformer_engine/pytorch/module.py:170]: But my point is that those pointers should already be correct, otherwise the result is wrong, right? Or am I missing something?

File: transformer_engine/pytorch/module.py [hunk 3]
```diff
@@ -993,6 +1031,11 @@ class LayerNormLinear(TransformerEngineBaseModule):
                              together with the output of the linear transformation.
                              Example use case: residual connection for transformer module is
                              taken post layernorm.
+    wandb_param_split : Tuple[str, ...], default = None
```
Comment (ptrendx) [inline transformer_engine/pytorch/module.py:1034]: What is the reason for that name?

Comment (ksivaman) [reply transformer_engine/pytorch/module.py:1034]: Just to make clear that it is for "weights and biases parameter split"

Comment (ptrendx) [reply transformer_engine/pytorch/module.py:1034]: Hmmm, if you did not tell me that I would think that we have some dependency on WandB :-P. Could we make it a different name?

Comment (ksivaman) [reply transformer_engine/pytorch/module.py:1034]: Makes sense

File: transformer_engine/pytorch/module.py (hunk 2 above)
Comment (ksivaman) [reply transformer_engine/pytorch/module.py:170]: In the general use case for `NoopCat`. If `params_split` occupy different memory region from `full_param_buffer`, then we would need to set them explicitly, right?

**Comment by ksivaman** [comment]
@ptrendx Instead of intercepting all the convert functions for parameters and buffers (there are a lot), which would be a bit flimsy, this is the solution I have come up with; i.e. if the parameter and buffer memory is different, overwrite the buffer using actual concatenation before calling `NoopCat`.

File: transformer_engine/pytorch/module.py [hunk 4]
```diff
@@ -572,6 +608,23 @@ def grad_output_preprocess(
...
+        """
+
+        assert hasattr(self, buffer_name), f"No buffer named {buffer_name}"
+        full_param_buffer = getattr(self, buffer_name)
+        split_size = full_param_buffer.shape[0] // len(params_split)
+        for i, p in enumerate(params_split):
+            if p.data.data_ptr() != full_param_buffer[i*split_size : (i+1)*split_size].data_ptr():
+                with torch.no_grad():
+                    setattr(self, buffer_name, torch.cat(params_split))
+                break
```
Comment (ptrendx) [inline transformer_engine/pytorch/module.py:624]: So just to be clear - this does not then "fix" the problem of the split params not being in a single buffer -> it will just perform the concat every time in that case, right?

Is it possible to update the params after this cat so that they point to the right place and on subsequent iteration concat is not needed anymore?

Comment (ksivaman) [reply transformer_engine/pytorch/module.py:624]: Good catch, forgot to update the params memory :)

Comment (ksivaman) [reply transformer_engine/pytorch/module.py:630]: /te-ci

**Review by ptrendx — APPROVED** [review]
LGTM
